### DISCUSSION TITLE
Python: User approval / Force function invocation in specific order

### DISCUSSION THREAD

**Discussion by WojciechTyczynski** [microsoft/semantic-kernel#9393] [Q&A] [answered]
Hi team,

I'm currently working with the `ChatCompletionAgent` in Python and I need to ensure that one function is always called before another, specifically to obtain user approval.

Here's the context: I have an app that uses the `ChatCompletionAgent` served as an API with a single endpoint (POST on_message). The agent is enhanced with a plugin called DataHandler, which consists of four functions: `load_data`, `edit_data`, `push_data`, and `approve_action`. My goal is to always invoke `approve_action` before `push_data` to get user approval.

I've looked into examples of how to obtain user approval using filters, but they don't seem to work in this case because serving the Agent as an API requires something to be returned to obtain extra user input.

**Comment by moonbox3** [comment, marked as answer]
Hi @WojciechTyczynski, thanks for your question -- and it's a very good one. As of right now, when simply using the agent framework, it can be a bit tricky to always get the repeatable nature of calling certain plugins in the desired order. You can try to use instructions to explain the order flow of how the agent should interact with the plugins it has available. But again, that isn't always repeatable.

We're working on getting in the Process Framework for Python as soon as possible. This will definitely help your scenario as you can specify that the proper KernelFunction (plugin) be run always before another step. Once the PR goes in, I will work on adding more samples in Python, including a sample on how to construct a process using agents.

For now, you can have a look at the process framework PR: https://github.com/microsoft/semantic-kernel/pull/9363. 

You can also have a look at a sample linear process that is included in the PR: https://github.com/microsoft/semantic-kernel/blob/d039a8417b34e0c862faacbcf320c27875762a31/python/samples/getting_started_with_processes/step01_processes.py

This should be in main very soon.
